zcash_client_backend/src/data_api/wallet/input_selection.rs

```diff
@@ -1039,20 +1039,37 @@ impl<DbT: InputSource> ShieldingSelector for GreedyInputSelector<DbT> {
         ParamsT: consensus::Parameters,
         ChangeT: ChangeStrategy<MetaSource = Self::InputSource>,
     {
-        let mut transparent_inputs: Vec<WalletTransparentOutput> = source_addrs
-            .iter()
-            .map(|taddr| {
-                wallet_db.get_spendable_transparent_outputs(
-                    taddr,
-                    target_height,
-                    confirmations_policy,
-                )
-            })
-            .collect::<Result<Vec<Vec<_>>, _>>()
-            .map_err(InputSelectorError::DataSource)?
-            .into_iter()
-            .flat_map(|v| v.into_iter())
-            .collect();
+        let (mut transparent_inputs, _, _) = source_addrs.iter().try_fold(
+            (
+                vec![],
+                BTreeSet::<TransparentAddress>::new(),
+                BTreeSet::<TransparentAddress>::new(),
+            ),
+            |(mut inputs, mut ephemeral_addrs, mut input_addrs), taddr| {
+                use transparent::keys::TransparentKeyScope;
+
+                let utxos = wallet_db
+                    .get_spendable_transparent_outputs(taddr, target_height, confirmations_policy)
+                    .map_err(InputSelectorError::DataSource)?;
+
+                ephemeral_addrs.extend(utxos.iter().filter_map(|utxo| {
+                    (utxo.recipient_key_scope() == Some(TransparentKeyScope::EPHEMERAL))
+                        .then_some(utxo.recipient_address())
+                }));
+                input_addrs.extend(utxos.iter().map(|utxo| utxo.recipient_address()));
+                inputs.extend(utxos.into_iter().map(|utxo| utxo.into_wallet_output()));
+
+                // Funds may be spent from at most one ephemeral address at a time. If there are no
+                // ephemeral addresses, we allow shielding from multiple transparent addresses.
+                if !ephemeral_addrs.is_empty() && input_addrs.len() > 1 {
```

Comment:
This doesn't take into account transparent recipient addresses belonging to this wallet. That is, arguably the policy should be that the union of source and recipient addresses should not include multiple transparent addresses belonging to this wallet, if at least one of them is ephemeral. Note that ZIP 320 satisfies this.